godot-core/src/engine/file.rs

```diff
@@ -0,0 +1,766 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use crate::builtin::{GodotString, PackedByteArray, PackedStringArray, Variant};
+use crate::engine::file_access::{CompressionMode, ModeFlags};
+use crate::gen::central::global::Error;
+use crate::gen::classes::FileAccess;
+use crate::obj::{Gd, Inherits};
+
+use std::cmp;
+use std::io::{BufRead, Error as IoError, ErrorKind, Read, Result as IoResult, Seek, Write};
+
+use super::RefCounted;
+
+/// Open file for reading or writing.
+///
+/// It's a convienience wrapper around [Gd]<[FileAccess]>, providing multiple
+/// both safety and quality of life upgrades over the inner type.
+///
+/// Files by default are alwasy opened with little-endian endianness, as most
+/// files are saved as such. To switch into big_endian_edianness, use
+/// `GFile.set_big_endian()`
+///  
+/// - exposes reading and writing capabilities of [FileAccess] in a safe way,
+/// returning [Result<T>](std::io::Result) instead of the `T` itself
+/// - makes the [FileAccess] exclusive to its instance, not allowing parallel
+/// reads and writes, which could introduce hard to track bugs by unpredicted cursor
+/// movement. Exclusivity also ensures that when the `GFile` moves out of scope,
+/// inner [FileAccess] does also, automatically closing the file. Alternatively,
+/// you can [drop()] the `GFile` to close the file manually.
+/// - implements useful Rust traits allowing better file processing, namely
+/// [Read], [BufRead], [Write], [Seek] and making it better integrated with
+/// various tools in Rust ecosystem (eg. serialization tools).
+///
+/// ## [ModeFlags]
+///
+/// Every constructor (`open_*` associated function) intakes a `flags`
+/// argument, opening the file for different types of operations. Regardless
+/// of provided `flags`, cursor is always positioned at the beginning of the
+/// file upon opening. To adjust its position use [Seek]-provided methods.
+///
+/// - `ModeFlags::Read` - opens file for read operations.
+/// - `ModeFlags::Write` - opens file for write operations. If the file doesn't
+/// exist at provided `path`, it is created. If it exists, it is truncated
+/// after the file is closed.
+/// - `ModeFlags::ReadWrite` - opens the file for read and write operations.
+/// The file is not truncated after closing.
+/// - `ModeFlags::WriteRead` - opens the file for read and write operations.
+/// If the file doesn't exist at provided `path`, it is created. If it exists,
+/// it is truncated after the file is closed .
+///
+/// ## Examples
+///
+/// ```no_run
+/// // Open file in write mode
+/// let my_file = GFile::open("user://save_game.sav".into(), ModeFlags::Write);
+/// // Write some `GodotString`` lines into it
+/// my_file.write_godot_line("This is my saved game").expect("Couldn't write first line");
+/// my_file.write_godot_line("I played for 5 minutes").expect("Couldn't write second line");
```

Comment:
All these sections could have empty line before `//` comments, to group related statements. In general, this might make things a bit more readable in doctests.

Also, double backtick after "GodotString" 🙂